sql-serializer/src/main/java/io/cdap/pipeline/sql/serializer/bigquery/BigQuerySQLSerializer.java

```diff
@@ -0,0 +1,247 @@
+    HashSet<Column> definedColumns = new HashSet<>();
+    if (isSubquery) {
+      b.append('(');
+    } else if (query.hasAlias()) {
+      throw new IllegalArgumentException("Cannot have an alias on the main outer query");
+    }
+    b.append("SELECT ");
+    for (int i = 0; i < query.getColumns().size(); i++) {
+      Column col = query.getColumns().get(i);
+      if (col.hasAlias()) {
+        definedColumns.add(col);
+      }
+      appendColumnSQL(b, col, true, false);
+      if (i < query.getColumns().size() - 1) {
+        b.append(", ");
+      }
+    }
+    b.append(" FROM ");
+    appendQueryableSQL(b, query.getQueryFrom());
+    if (query.getFilter() != null) {
+      b.append(" WHERE ");
+      appendFilterSQL(b, query.getFilter(), definedColumns);
+    }
+    if (isSubquery) {
+      b.append(')');
+      if (query.hasAlias()) {
+        b.append(" AS ");
+        b.append(query.getAlias());
+      }
+    } else {
+      b.append(';');
+    }
+    return b;
+  }
+
+  private StringBuilder appendColumnSQL(StringBuilder b, Column c,
+                                        boolean defineAliasIfExists, boolean useAliasIfExists) {
+    if (defineAliasIfExists && useAliasIfExists) {
+      throw new IllegalArgumentException("Unable to use and define an alias at the same time");
+    }
+    if (c.hasSource() && c.getSource().hasAlias()) {
+      b.append(c.getSource().getAlias());
+      b.append('.');
+    }
+    if (c.hasAlias() && useAliasIfExists) {
+      b.append(c.getAlias());
+    } else {
+      b.append(c.getName());
+    }
+    if (c.hasAlias() && defineAliasIfExists) {
+      b.append(" AS ");
+      b.append(c.getAlias());
+    }
+    return b;
+  }
+
+  private StringBuilder appendTableSQL(StringBuilder b, Table t, boolean defineAliasIfExists) {
+    if (!t.hasAlias() || defineAliasIfExists) {
+      b.append('`');
+      if (t.hasProject()) {
+        b.append(t.getProjectName());
+        b.append('.');
+      }
+      if (t.hasDatabase()) {
+        b.append(t.getDatabaseName());
+        b.append('.');
+      }
+      b.append(t.getTableName());
+      b.append('`');
+      if (t.hasAlias() && defineAliasIfExists) {
+        b.append(" AS ");
+        b.append(t.getAlias());
+      }
+    } else {
+      b.append('`');
+      b.append(t.getAlias());
+      b.append('`');
+    }
+    return b;
+  }
+
+  private StringBuilder appendQueryableSQL(StringBuilder b, Queryable s) {
+    QueryableType type = s.getSourceType();
+    if (type.equals(QueryableType.QUERY)) {
+      return appendQuerySQL(b, (StructuredQuery) s, true);
+    } else if (type.equals(QueryableType.TABLE)) {
+      return appendTableSQL(b, (Table) s, true);
+    } else {
+      throw new IllegalArgumentException("Invalid readable source type");
+    }
+  }
+
+  private StringBuilder appendFilterSQL(StringBuilder b, Filter f, Set<Column> definedColumns) {
```

Comment:
This class could use more javadocs. Can you add one to each method that contains an example of what each method is generating

Reply:
Added some javadocs and examples